qiskit/aqua/algorithms/phase_estimators/hamiltonian_pe_result.py

```diff
@@ -0,0 +1,79 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Result from running HamiltonianPE"""
+
+
+from typing import Dict, List, Union
+from qiskit.result import Result
+from .phase_estimator_result import PhaseEstimatorResult
+from .phase_estimation_scale import PhaseEstimationScale
+
+
+class HamiltonianPEResult(PhaseEstimatorResult):
+    """Store and manipulate results from running `HamiltonianPE`.
+
+    This API of this class is nearly the same as `PhaseEstimatorResult`, differing only in
+    the presence of an additional keyword argument in the methods. If `scaled`
+    is `False`, then the phases are not translated and scaled to recover the
+    eigenvalues of the Hamiltonian. Instead `phi` in `[0, 1)` is returned,
+    as is the case when then unitary is not derived from a Hamiltonian.
+    """
+    def __init__(self, num_evaluation_qubits: int, circuit_result: Result,
+                 phase_estimation_scale: PhaseEstimationScale,
+                 phase_array: List = None, phase_dict: Dict = None) -> None:
+
+        self._phase_estimation_scale = phase_estimation_scale
+
+        super().__init__(num_evaluation_qubits, circuit_result, phase_array, phase_dict)
+
+    # pylint: disable=arguments-differ
+    def filter_phases(self, cutoff: float = 0.0, scaled: bool = True,  # type: ignore
+                      as_float: bool = True) -> Union[Dict, List]:  # returns a dict
+        """Filter phases as does `PhaseEstimatorResult.filter_phases`, with
+        the addition that `phi` is shifted and translated to return eigenvalues
+        of the Hamiltonian.
+
+        Args:
+            cutoff (float): Minimum weight of number of counts required to keep a bit string.
+                          The default value is `0.0`.
+            scaled (bool): If False, return `phi` in `[0, 1)` rather than the eigenvalues of
+                         the Hamiltonian.
+            as_float (bool): If `True`, returned keys are floats in `[0.0, 1)`. If `False`
+                      returned keys are bit strings.
+
+        Returns:
+              A dict of filtered phases.
```

Comment:
Is this a Dict or a List - the return type says it could be either but the comment there and here just say dict

Reply:
In this
```
 -> Union[Dict, List]:  # returns a dict
```
 the `Union` is to silence the mypy tool. And the comment is to tell the reader that it's really a dict.  mypy's inference engine apparently doesn't work very well. There's probably another way around this.

Reply:
I think it has more to do with `return self._phase_estimation_scale.scale_phases(phases)` returning a Union[List, Dict] so I would have thought mypy was correct in this case and the commenting is wrong.

Reply:
Clearly `phases = super().filter_phases(cutoff, as_float=as_float)` binds `phases` to a `dict`.
Then if `scaled` is `True`, we have `return self._phase_estimation_scale.scale_phases(phases)`.
Note that `_phase_estimation_scale.scale_phases(phases)` returns a `list` if passed a `list` and returns a `dict` if passed a `dict`. So you can infer that `hamiltonian_pe_result.filter_phases` always returns a `dict`.

This is an easy inference for a language that supports inferring output type from input types. So, I suspect you are correct that it is not a failure of mypy, but rather this kind of inference was never intended to be part of the mypy system's semantics. Perhaps for a good reason.

If you glance at code with mypy annotations, they appear complicated and occupy a lot of your attention, giving the impression that you are dealing with a language that takes types seriously. But, to me, they see more of a hindrance.
Still, some big projects use mypy. All of this tells me I just don't understand it yet, and I need to spend some time reading about mypy and how to deal with situations like this one, which are probably common.

Reply:
If you are assured that filter phases always returns a Dict, given you supply a Dict, then you can cast to that which should allow you to get rid of the Union.

Reply:
> Casts are used to silence spurious type checker warnings and give the type checker a little help when it can’t quite understand what is going on.

After looking into it, it appears this is the officially-sanctioned solution. Best, not to try to make sense of any of this. Just make the change and move on.

I find the `filter_phases` method a bit  ill-designed anyway. So maybe it will be moot in the end. But, that's another question.